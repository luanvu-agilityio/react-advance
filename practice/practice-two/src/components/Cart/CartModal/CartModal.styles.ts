import styled from 'styled-components'

export const ModalOverlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 1000;
  display: flex;
  justify-content: flex-end;
  backdrop-filter: blur(5px);
`

export const ModalContent = styled.div`
  width: 400px;
  max-height: 912px;
  background-color: white;
  height: 100%;
  padding: 24px 16px;
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  box-shadow: 0px 2px 32px 0px #00000026;
`

export const ModalHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 24px;
`

export const ModalTitle = styled.h2`
  font-size: 26px;
  font-weight: var(--font-weight-semibold);
  margin: 0;
`

export const CloseButton = styled.button`
  display: flex;
  align-items: center;
  background: none;
  border: none;
  font-size: 24px;
  cursor: pointer;
  padding: 0;

  span {
    font-family: var(--font-family-secondary);
    font-size: 12px;
    font-weight: var(--font-weight-regular);
    color: var(--black-shade-3);
  }
`

export const CartItemList = styled.div`
  flex: 1;
  overflow-y: auto;
  margin-bottom: 24px;
  display: flex;
  flex-direction: column;
  gap: 32px;
`

export const CartFooter = styled.div`
  margin-top: auto;
`

export const Subtotal = styled.div`
  display: flex;
  flex-direction: column;
  padding: 24px 0;
  font-size: 26px;
  font-weight: var(--font-weight-semibold);
`

export const ButtonsContainer = styled.div`
  display: flex;
  justify-content: space-between;
  padding-top: 16px;
  border-top: 1px solid var(--black-shade-4);
`

export const Button = styled.button`
  border-radius: 12px;
  font-weight: var(--font-weight-medium);
  cursor: pointer;
  text-align: center;
`

export const EmptyCartMessage = styled.div`
  text-align: center;
  padding: 48px 0;
  color: var(--black-shade-2);
`
