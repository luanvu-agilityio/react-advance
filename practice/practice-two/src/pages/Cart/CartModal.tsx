import { useEffect } from 'react'
import styled from 'styled-components'
import { useCart } from '@contexts/CartContext'
import CartItem from './CartItem'
import { useNavigate } from 'react-router-dom'

const ModalOverlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 1000;
  display: flex;
  justify-content: flex-end;
`

const ModalContent = styled.div`
  width: 400px;
  max-height: 912px;
  background-color: white;
  height: 100%;
  padding: 24px 16px;
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  box-shadow: 0px 2px 32px 0px #00000026;
`

const ModalHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 24px;
`

const ModalTitle = styled.h2`
  font-size: 26px;
  font-weight: var(--font-weight-semibold);
  margin: 0;
`

const CloseButton = styled.button`
  display: flex;
  align-items: center;
  background: none;
  border: none;
  font-size: 24px;
  cursor: pointer;
  padding: 0;

  span {
    font-family: var(--font-family-secondary);
    font-size: 12px;
    font-weight: var(--font-weight-regular);
    color: var(--black-shade-3);
  }
`

const CartItemList = styled.div`
  flex: 1;
  overflow-y: auto;
  margin-bottom: 24px;
  display: flex;
  flex-direction: column;
  gap: 32px;
`

const CartFooter = styled.div`
  margin-top: auto;
`

const Subtotal = styled.div`
  display: flex;
  flex-direction: column;
  padding: 24px 0;
  font-size: 26px;
  font-weight: var(--font-weight-semibold);
`

const ButtonsContainer = styled.div`
  display: flex;
  justify-content: space-between;
  padding-top: 16px;
  border-top: 1px solid var(--black-shade-4);
`

const Button = styled.button`
  border-radius: 12px;
  font-weight: var(--font-weight-medium);
  cursor: pointer;
  text-align: center;
`

const PrimaryButton = styled(Button)`
  background-color: var(--green-color-default);
  color: white;
  border: none;
  padding: 12px 16px;
  font-size: 15px;
  height: 48px;
  font-weight: var(--font-weight-bold);
`

const SecondaryButton = styled(Button)`
  background-color: white;
  color: var(--black-color-default);
  border: none;
  font-size: 15px;
  font-weight: var(--font-weight-bold);
`

const EmptyCartMessage = styled.div`
  text-align: center;
  padding: 48px 0;
  color: var(--black-shade-2);
`

const CartModal = () => {
  const { items, removeItem, updateItem, isOpen, closeCart, subtotal } =
    useCart()
  const navigate = useNavigate()
  // Close modal when Escape key is pressed
  useEffect(() => {
    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        closeCart()
      }
    }

    window.addEventListener('keydown', handleEscape)
    return () => {
      window.removeEventListener('keydown', handleEscape)
    }
  }, [closeCart])

  // Prevent scrolling of body when modal is open
  useEffect(() => {
    if (isOpen) {
      document.body.style.overflow = 'hidden'
    } else {
      document.body.style.overflow = 'unset'
    }
    return () => {
      document.body.style.overflow = 'unset'
    }
  }, [isOpen])

  if (!isOpen) return null

  // Handle quantity change for specific item
  const handleQuantityChange = (id: number, quantity: number) => {
    updateItem(id, { quantity })
  }

  // Handle buy unit change for specific item
  const handleUnitChange = (id: number, buyUnit: string) => {
    updateItem(id, { buyUnit })
  }
  const handleCheckout = () => {
    closeCart() // Close the cart modal first
    navigate('/checkout') // Navigate to checkout page
  }
  return (
    <ModalOverlay onClick={closeCart}>
      <ModalContent onClick={(e) => e.stopPropagation()}>
        <ModalHeader>
          <ModalTitle>Shopping cart</ModalTitle>
          <CloseButton onClick={closeCart}>
            <span>Close</span>Ã—
          </CloseButton>
        </ModalHeader>

        <CartItemList>
          {items.length === 0 ? (
            <EmptyCartMessage>
              <p>Your cart is empty</p>
              <SecondaryButton onClick={closeCart}>
                Continue shopping
              </SecondaryButton>
            </EmptyCartMessage>
          ) : (
            items.map((item) => (
              <CartItem
                key={`${item.id}-${item.buyUnit}`}
                item={item}
                onQuantityChange={handleQuantityChange}
                onUnitChange={handleUnitChange}
                onRemove={removeItem}
              />
            ))
          )}
        </CartItemList>

        {items.length > 0 && (
          <CartFooter>
            <Subtotal>
              <span style={{ fontSize: '12px' }}>Subtotal</span>
              <span>{subtotal.toFixed(2)} USD</span>
            </Subtotal>
            <ButtonsContainer>
              <SecondaryButton onClick={closeCart}>
                Continue shopping
              </SecondaryButton>
              <PrimaryButton onClick={handleCheckout}>
                Go to Checkout
              </PrimaryButton>
            </ButtonsContainer>
          </CartFooter>
        )}
      </ModalContent>
    </ModalOverlay>
  )
}

export default CartModal
