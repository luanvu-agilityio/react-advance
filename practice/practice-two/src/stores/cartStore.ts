import type { Product } from 'types/Product'
import type { CartItem } from 'types/cart-items'
import { create } from 'zustand'
import { persist } from 'zustand/middleware'

interface CartState {
  // State
  items: CartItem[]
  isOpen: boolean

  // Actions
  addItem: (
    product:
      | Product
      | (Partial<Product> & { id: number; price: number; title: string }),
    quantity: number,
    buyUnit: string
  ) => void
  removeItem: (id: number) => void
  updateItem: (id: number, updatedItem: Partial<CartItem>) => void
  clearCart: () => void
  openCart: () => void
  closeCart: () => void
  updateQuantity: (id: number, quantity: number) => void
  updateUnit: (id: number, buyUnit: string) => void

  // Computed values
  getSubtotal: () => number
  getTax: () => number
  getShippingCost: (method: string) => number
  getTotal: (shippingMethod: string) => number
}

export const useCartStore = create<CartState>()(
  persist(
    (set, get) => ({
      items: [],
      isOpen: false,

      addItem: (product, quantity, buyUnit) => {
        set((state) => {
          const existingItemIndex = state.items.findIndex(
            (i) => i.id === product.id && i.buyUnit === buyUnit
          )
          if (existingItemIndex !== -1) {
            const newItems = [...state.items]
            newItems[existingItemIndex].quantity += quantity
            return { items: newItems, isOpen: true }
          } else {
            const newItem = {
              id: product.id,
              title: product.title,
              price: product.price,
              quantity,
              buyUnit,
              imageUrl: product.imageUrl,
              farm: product.farm,
              freshness: product.freshness,
              rating: product.rating,
            }
            return {
              items: [...state.items, newItem],
              isOpen: true,
            }
          }
        })
      },

      removeItem: (id) =>
        set((state) => ({
          items: state.items.filter((item) => item.id !== id),
        })),

      updateItem: (id, updates) =>
        set((state) => ({
          items: state.items.map((item) =>
            item.id === id ? { ...item, ...updates } : item
          ),
        })),

      clearCart: () => set({ items: [] }),
      openCart: () => set({ isOpen: true }),
      closeCart: () => set({ isOpen: false }),

      getSubtotal: () => {
        const { items } = get()
        return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
      },

      getTax: () => get().getSubtotal() * 0.17, //  17% tax rate

      getShippingCost: (method) => {
        switch (method) {
          case 'fedex':
            return 32
          case 'dhl':
            return 15
          default:
            return 15
        }
      },

      getTotal: (shippingMethod) => {
        const subtotal = get().getSubtotal()
        const tax = get().getTax()
        const shippingCost = get().getShippingCost(shippingMethod)
        return subtotal + tax + shippingCost
      },

      updateQuantity: (id, quantity) =>
        set((state) => ({
          items: state.items.map((item) =>
            item.id === id ? { ...item, quantity } : item
          ),
        })),

      updateUnit: (id, buyUnit) =>
        set((state) => ({
          items: state.items.map((item) =>
            item.id === id ? { ...item, buyUnit } : item
          ),
        })),
    }),
    {
      name: 'cart-storage',
    }
  )
)
