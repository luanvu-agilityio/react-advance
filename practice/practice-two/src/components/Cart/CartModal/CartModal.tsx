import { useEffect, type MouseEvent } from 'react'
import { useNavigate } from 'react-router-dom'
import { useCartStore } from '@stores/cartStore'
import {
  ButtonsContainer,
  CartFooter,
  CartItemList,
  CloseButton,
  EmptyCartMessage,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  ModalTitle,
  Subtotal,
} from './CartModal.styles'
import CartItem from '../CartItem/CartItem'
import { useRemoveFromCart } from '@hooks/useCartMutation'
import { Button } from '@radix-ui/themes'

const CartModal = () => {
  const { items, updateItem, isOpen, closeCart, getSubtotal } = useCartStore()
  const navigate = useNavigate()

  const { mutate: removeCartItem } = useRemoveFromCart()

  // Close modal when Escape key is pressed
  useEffect(() => {
    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        closeCart()
      }
    }

    window.addEventListener('keydown', handleEscape)
    return () => {
      window.removeEventListener('keydown', handleEscape)
    }
  }, [closeCart])

  // Prevent scrolling of body when modal is open
  useEffect(() => {
    if (isOpen) {
      document.body.style.overflow = 'hidden'
    } else {
      document.body.style.overflow = 'unset'
    }
    return () => {
      document.body.style.overflow = 'unset'
    }
  }, [isOpen])

  if (!isOpen) return null

  // Handle quantity change for specific item
  const handleQuantityChange = (id: number, quantity: number) => {
    updateItem(id, { quantity })
  }

  // Handle buy unit change for specific item
  const handleUnitChange = (id: number, buyUnit: string) => {
    updateItem(id, { buyUnit })
  }

  const handleCheckout = () => {
    closeCart() // Close the cart modal first
    navigate('/checkout') // Navigate to checkout page
  }

  const handleRemoveItem = (id: number) => {
    const itemToRemove = items.find((item) => item.id === id)
    if (itemToRemove) {
      removeCartItem({
        id,
        title: itemToRemove.title,
      })
    }
  }
  return (
    <ModalOverlay data-testid="modal-overlay" onClick={closeCart}>
      <ModalContent onClick={(e: MouseEvent) => e.stopPropagation()}>
        <ModalHeader>
          <ModalTitle>Shopping cart</ModalTitle>
          <CloseButton onClick={closeCart}>
            <span>Close</span>Ã—
          </CloseButton>
        </ModalHeader>

        <CartItemList>
          {items.length === 0 ? (
            <EmptyCartMessage>
              <p>Your cart is empty</p>
              <Button variant="soft" onClick={closeCart}>
                Continue shopping
              </Button>
            </EmptyCartMessage>
          ) : (
            items.map((item) => (
              <CartItem
                key={`${item.id}-${item.buyUnit}`}
                item={item}
                onQuantityChange={handleQuantityChange}
                onUnitChange={handleUnitChange}
                onRemove={handleRemoveItem}
              />
            ))
          )}
        </CartItemList>

        {items.length > 0 && (
          <CartFooter>
            <Subtotal>
              <span style={{ fontSize: '12px' }}>Subtotal</span>
              <span>{getSubtotal().toFixed(2)} USD</span>
            </Subtotal>
            <ButtonsContainer>
              <Button variant="soft" onClick={closeCart}>
                Continue shopping
              </Button>
              <Button
                variant="solid"
                onClick={handleCheckout}
                style={{ padding: '12px 16px' }}
              >
                Go to Checkout
              </Button>
            </ButtonsContainer>
          </CartFooter>
        )}
      </ModalContent>
    </ModalOverlay>
  )
}

export default CartModal
