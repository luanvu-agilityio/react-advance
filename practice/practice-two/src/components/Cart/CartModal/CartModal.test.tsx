import { render, screen, fireEvent, act } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import CartModal from './CartModal'
import { useCartStore } from '@stores/cartStore'
import type { ReactElement, ReactNode } from 'react'

// Mock the router navigation
const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
  // Only mock the components we actually use
  MemoryRouter: ({ children }: { children: ReactNode }) => (
    <div data-testid="memory-router">{children}</div>
  ),
  Link: ({ to, children }: { to: string; children: ReactNode }) => (
    <a href={to} data-testid="router-link">
      {children}
    </a>
  ),
}))

// Mock the Zustand store
jest.mock('@stores/cartStore', () => ({
  useCartStore: jest.fn(),
}))

// Mock the toast context
jest.mock('@contexts/ToastContext', () => ({
  useToast: () => ({
    toast: jest.fn(),
  }),
}))

// Mock the cart mutation hook
const mockRemoveCartMutation = jest.fn()
jest.mock('@hooks/useCartMutation', () => ({
  useRemoveFromCart: () => ({
    mutate: mockRemoveCartMutation,
    isPending: false,
  }),
}))

// Helper function to wrap component in router
const renderWithRouter = (component: ReactElement) => {
  return render(<MemoryRouter>{component}</MemoryRouter>)
}

describe('CartModal', () => {
  // Common mock values
  const mockCloseCart = jest.fn()
  const mockUpdateItem = jest.fn()
  const mockGetSubtotal = jest.fn(() => 42.99)
  const mockItems = [
    {
      id: 1,
      title: 'Test Product 1',
      price: 12.99,
      quantity: 2,
      buyUnit: 'kg',
      image: '/img.jpg',
    },
    {
      id: 2,
      title: 'Test Product 2',
      price: 17.01,
      quantity: 1,
      buyUnit: 'box',
      image: '/img2.jpg',
    },
  ]

  // Reset mocks between tests
  beforeEach(() => {
    jest.clearAllMocks()
  })

  // SNAPSHOT TESTS FOR RENDERING

  it('matches snapshot when modal is closed', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: false,
      items: [],
      closeCart: mockCloseCart,
    })

    const { container } = renderWithRouter(<CartModal />)
    expect(container).toMatchSnapshot()
  })

  it('matches snapshot with empty cart', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: [],
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    const { container } = renderWithRouter(<CartModal />)
    expect(container).toMatchSnapshot()
  })

  it('matches snapshot with items in cart', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: mockItems,
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    const { container } = renderWithRouter(<CartModal />)
    expect(container).toMatchSnapshot()
  })

  // INTERACTIVE BEHAVIOR TESTS

  it('closes cart when clicking outside modal', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: [],
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    const { container } = renderWithRouter(<CartModal />)
    const overlay =
      screen.queryByTestId('modal-overlay') ||
      container.querySelector('.backdrop, .overlay, .dialog-overlay')

    if (!overlay) {
      throw new Error(
        'Could not find modal overlay - check the component structure'
      )
    }

    fireEvent.click(overlay)
    expect(mockCloseCart).toHaveBeenCalledTimes(1)
  })

  it('does not close cart when clicking inside modal content', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: [],
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    renderWithRouter(<CartModal />)
    const modalContent = screen.getByText('Shopping cart').closest('div')
    if (modalContent) {
      fireEvent.click(modalContent)
    }
    expect(mockCloseCart).not.toHaveBeenCalled()
  })

  it('closes cart when pressing Escape key', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: [],
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    renderWithRouter(<CartModal />)
    act(() => {
      fireEvent.keyDown(document, { key: 'Escape' })
    })
    expect(mockCloseCart).toHaveBeenCalledTimes(1)
  })

  it('navigates to checkout when clicking checkout button', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: mockItems,
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    renderWithRouter(<CartModal />)
    const checkoutButton = screen.getByText('Go to Checkout')
    fireEvent.click(checkoutButton)
    expect(mockCloseCart).toHaveBeenCalledTimes(1)
    expect(mockNavigate).toHaveBeenCalledWith('/checkout')
  })

  it('calls closeCart when clicking continue shopping button', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: [],
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    renderWithRouter(<CartModal />)
    const continueButton = screen.getByText('Continue shopping')
    fireEvent.click(continueButton)
    expect(mockCloseCart).toHaveBeenCalledTimes(1)
  })

  // NEW TESTS FOR REMOVE FUNCTIONALITY WITH TOAST

  it('calls removeCartItem mutation when removing an item', async () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: mockItems,
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    // Render the component
    renderWithRouter(<CartModal />)

    // Find the CartItem component - we'll need to add a data-testid to the remove button in CartItem
    // For now, let's assume there's a button with text "Remove"
    const removeButtons = screen.getAllByText('Remove')
    fireEvent.click(removeButtons[0]) // Remove the first item

    // Check that the mutation was called with correct arguments
    expect(mockRemoveCartMutation).toHaveBeenCalledWith({
      id: 1, // ID of the first item
      title: 'Test Product 1', // Title of the first item
    })
  })

  it('handles quantity change for cart items', () => {
    ;(useCartStore as unknown as jest.Mock).mockReturnValue({
      isOpen: true,
      items: mockItems,
      closeCart: mockCloseCart,
      updateItem: mockUpdateItem,
      getSubtotal: mockGetSubtotal,
    })

    renderWithRouter(<CartModal />)

    // Since we can't easily test BuyingUnit component behavior here,
    // we can verify the update functions are passed to CartItem

    // First, make sure there are cart items rendered
    expect(screen.getAllByText(/Test Product/)).toHaveLength(2)

    // Verify that updateItem is not called yet
    expect(mockUpdateItem).not.toHaveBeenCalled()
  })
})
