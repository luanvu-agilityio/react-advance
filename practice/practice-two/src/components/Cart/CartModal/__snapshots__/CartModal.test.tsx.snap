// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartModal matches snapshot when modal is closed 1`] = `
<div>
  <div
    data-testid="memory-router"
  />
</div>
`;

exports[`CartModal matches snapshot with empty cart 1`] = `
<div>
  <div
    data-testid="memory-router"
  >
    <div
      class="sc-Qotzb docwpJ"
      data-testid="modal-overlay"
    >
      <div
        class="sc-fYsHOw cvEzIP"
      >
        <div
          class="sc-dKREkF fFaeCn"
        >
          <h2
            class="sc-fWnslK bGMzDG"
          >
            Shopping cart
          </h2>
          <button
            class="sc-dIMoHT feVYAI"
          >
            <span>
              Close
            </span>
            ×
          </button>
        </div>
        <div
          class="sc-iQQCXo kXdUuw"
        >
          <div
            class="sc-cXawGu jlGxAR"
          >
            <p>
              Your cart is empty
            </p>
            <button
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
              data-accent-color=""
            >
              Continue shopping
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CartModal matches snapshot with items in cart 1`] = `
<div>
  <div
    data-testid="memory-router"
  >
    <div
      class="sc-Qotzb docwpJ"
      data-testid="modal-overlay"
    >
      <div
        class="sc-fYsHOw cvEzIP"
      >
        <div
          class="sc-dKREkF fFaeCn"
        >
          <h2
            class="sc-fWnslK bGMzDG"
          >
            Shopping cart
          </h2>
          <button
            class="sc-dIMoHT feVYAI"
          >
            <span>
              Close
            </span>
            ×
          </button>
        </div>
        <div
          class="sc-iQQCXo kXdUuw"
        >
          <div
            class="sc-fjUQFl iLFDDX"
          >
            <div
              style="display: flex; flex-direction: column; gap: 12px;"
            >
              <div
                class="sc-guGTOK hXwodv"
              >
                <img
                  alt="Test Product 1"
                />
              </div>
              <div
                class="sc-brzPDJ lnlnJc"
              >
                <button
                  class="sc-iMGFoU hWdhkH"
                >
                  <img
                    alt="Wishlist"
                    src="https://res.cloudinary.com/ds82onf5q/image/upload/v1748372436/wishlist_cltlcf.svg"
                  />
                  Wishlist
                </button>
                <button
                  class="sc-iMGFoU hWdhkH"
                >
                  <img
                    alt="Compare"
                    src="https://res.cloudinary.com/ds82onf5q/image/upload/v1748372430/compare_oeiv1m.svg"
                  />
                  Compare
                </button>
                <button
                  class="sc-iMGFoU hWdhkH"
                >
                  <img
                    alt="Remove"
                    src="https://res.cloudinary.com/ds82onf5q/image/upload/v1748372432/remove_gbwco0.svg"
                  />
                  Remove
                </button>
              </div>
            </div>
            <div
              class="sc-hZARmv hoQJX"
            >
              <div
                style="display: flex; flex-direction: column; gap: 4px;"
              >
                <h3
                  class="sc-dKKIkQ gNsqzZ"
                >
                  Test Product 1
                </h3>
                <div
                  style="display: flex; flex-direction: column; gap: 4px;"
                >
                  <div
                    class="sc-fpEFIB kfSVtb"
                  >
                    <span
                      class="sc-gAqISa gIMhYd"
                    >
                      Farm:
                    </span>
                    <span
                      class="sc-ifkGpL gUyMWZ"
                    />
                  </div>
                  <div
                    class="sc-fpEFIB kfSVtb"
                  >
                    <span
                      class="sc-gAqISa gIMhYd"
                    >
                      Freshness:
                    </span>
                    <span
                      class="sc-ifkGpL gUyMWZ"
                    />
                  </div>
                </div>
                <div
                  class="sc-jJAtPt iedezp"
                >
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                </div>
              </div>
              <div
                class="sc-gGarWV fjipvp"
              >
                <div
                  class="sc-fbguzk dTXkUE"
                >
                  12.99
                   USD
                </div>
                <div
                  class="sc-jScdur jrYsKT"
                >
                  <input
                    aria-label="Quantity"
                    class="sc-dcKlJK fXuBcx"
                    min="1"
                    type="number"
                    value="2"
                  />
                  <div
                    class="sc-jPkiSJ fSeUfZ"
                  />
                  <div
                    class="sc-cZSric jiVTkp"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      class="sc-lixPIL iooIYb"
                    >
                      kg
                      <svg
                        aria-hidden="true"
                        class="lucide lucide-chevron-down"
                        fill="none"
                        height="12"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="12"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="m6 9 6 6 6-6"
                        />
                      </svg>
                    </button>
                    <ul
                      class="sc-etfXYe chGDAh"
                      data-testid="dropdown-list"
                    >
                      <li
                        aria-selected="false"
                        class="sc-dFaThA goHwLH"
                      >
                        pcs
                      </li>
                      <li
                        aria-selected="false"
                        class="sc-dFaThA goHwLH"
                      >
                        kgs
                      </li>
                      <li
                        aria-selected="false"
                        class="sc-dFaThA goHwLH"
                      >
                        box
                      </li>
                      <li
                        aria-selected="false"
                        class="sc-dFaThA goHwLH"
                      >
                        pack
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-fjUQFl iLFDDX"
          >
            <div
              style="display: flex; flex-direction: column; gap: 12px;"
            >
              <div
                class="sc-guGTOK hXwodv"
              >
                <img
                  alt="Test Product 2"
                />
              </div>
              <div
                class="sc-brzPDJ lnlnJc"
              >
                <button
                  class="sc-iMGFoU hWdhkH"
                >
                  <img
                    alt="Wishlist"
                    src="https://res.cloudinary.com/ds82onf5q/image/upload/v1748372436/wishlist_cltlcf.svg"
                  />
                  Wishlist
                </button>
                <button
                  class="sc-iMGFoU hWdhkH"
                >
                  <img
                    alt="Compare"
                    src="https://res.cloudinary.com/ds82onf5q/image/upload/v1748372430/compare_oeiv1m.svg"
                  />
                  Compare
                </button>
                <button
                  class="sc-iMGFoU hWdhkH"
                >
                  <img
                    alt="Remove"
                    src="https://res.cloudinary.com/ds82onf5q/image/upload/v1748372432/remove_gbwco0.svg"
                  />
                  Remove
                </button>
              </div>
            </div>
            <div
              class="sc-hZARmv hoQJX"
            >
              <div
                style="display: flex; flex-direction: column; gap: 4px;"
              >
                <h3
                  class="sc-dKKIkQ gNsqzZ"
                >
                  Test Product 2
                </h3>
                <div
                  style="display: flex; flex-direction: column; gap: 4px;"
                >
                  <div
                    class="sc-fpEFIB kfSVtb"
                  >
                    <span
                      class="sc-gAqISa gIMhYd"
                    >
                      Farm:
                    </span>
                    <span
                      class="sc-ifkGpL gUyMWZ"
                    />
                  </div>
                  <div
                    class="sc-fpEFIB kfSVtb"
                  >
                    <span
                      class="sc-gAqISa gIMhYd"
                    >
                      Freshness:
                    </span>
                    <span
                      class="sc-ifkGpL gUyMWZ"
                    />
                  </div>
                </div>
                <div
                  class="sc-jJAtPt iedezp"
                >
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                  <span
                    class="sc-cUiCeM bRWKAv"
                  >
                    ★
                  </span>
                </div>
              </div>
              <div
                class="sc-gGarWV fjipvp"
              >
                <div
                  class="sc-fbguzk dTXkUE"
                >
                  17.01
                   USD
                </div>
                <div
                  class="sc-jScdur jrYsKT"
                >
                  <input
                    aria-label="Quantity"
                    class="sc-dcKlJK fXuBcx"
                    min="1"
                    type="number"
                    value="1"
                  />
                  <div
                    class="sc-jPkiSJ fSeUfZ"
                  />
                  <div
                    class="sc-cZSric jiVTkp"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      class="sc-lixPIL iooIYb"
                    >
                      box
                      <svg
                        aria-hidden="true"
                        class="lucide lucide-chevron-down"
                        fill="none"
                        height="12"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="12"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="m6 9 6 6 6-6"
                        />
                      </svg>
                    </button>
                    <ul
                      class="sc-etfXYe chGDAh"
                      data-testid="dropdown-list"
                    >
                      <li
                        aria-selected="false"
                        class="sc-dFaThA goHwLH"
                      >
                        pcs
                      </li>
                      <li
                        aria-selected="false"
                        class="sc-dFaThA goHwLH"
                      >
                        kgs
                      </li>
                      <li
                        aria-selected="true"
                        class="sc-dFaThA goHwLH"
                      >
                        box
                      </li>
                      <li
                        aria-selected="false"
                        class="sc-dFaThA goHwLH"
                      >
                        pack
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gDpztx eYGFlE"
        >
          <div
            class="sc-kpOvIu ckGWBO"
          >
            <span
              style="font-size: 12px;"
            >
              Subtotal
            </span>
            <span>
              42.99
               USD
            </span>
          </div>
          <div
            class="sc-icnseD ga-Dmrh"
          >
            <button
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
              data-accent-color=""
            >
              Continue shopping
            </button>
            <button
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
              data-accent-color=""
              style="padding: 12px 16px;"
            >
              Go to Checkout
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
